codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/DefaultEndpointProviderGenerator.kt

```diff
@@ -185,7 +204,7 @@ class DefaultEndpointProviderGenerator(
                     }
                 }
 
-                if (rule.endpoint.properties.isNotEmpty()) {
+                if (rule.endpoint.properties.isNotEmpty() || needAdditionalEndpointProperties) {
```

Comment:
suggestion: `rule.endpoint.properties` are mutable, so instead of creating this new boolean / tracking extra state, you can just add the `accountIdBasedEndpoint` and `serviceEndpointOverride` parameters to the endpoint params before entering this `if` statement.

Something like this:
```kt
rule.endpoint.properties[Identifier.of(RuntimeTypes.Core.BusinessMetrics.accountIdBasedEndPointAccountId.name)] = Literal.of("params.accountId")
rule.endpoint.properties[Identifier.of(RuntimeTypes.Core.BusinessMetrics.serviceEndpointOverride.name)] = Literal.booleanLiteral(true)

if (rule.endpoint.properties.isNotEmpty() { 
    ...
} 
```